fineract-loan/src/main/java/org/apache/fineract/portfolio/interestpauses/service/InterestPauseReadPlatformServiceImpl.java

```diff
@@ -0,0 +1,151 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class InterestPauseReadPlatformServiceImpl implements InterestPauseReadPlatformService {
+
+    private final PlatformSecurityContext context;
+    private final LoanTermVariationsRepository loanTermVariationsRepository;
+    private final LoanRepository loanRepository;
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<InterestPauseResponseDto> retrieveInterestPauses(Long loanId) {
+        this.context.authenticatedUser();
+
+        List<LoanTermVariationsData> variations = this.loanTermVariationsRepository.findLoanTermVariationsByLoanIdAndTermType(loanId,
+                LoanTermVariationType.INTEREST_PAUSE.getValue());
+
+        return mapToInterestPauseResponse(variations);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<InterestPauseResponseDto> retrieveInterestPausesByExternalId(String loanExternalId) {
+        this.context.authenticatedUser();
+
+        List<LoanTermVariationsData> variations = this.loanTermVariationsRepository.findLoanTermVariationsByExternalLoanIdAndTermType(
+                new ExternalId(loanExternalId), LoanTermVariationType.INTEREST_PAUSE.getValue());
+
+        return mapToInterestPauseResponse(variations);
+    }
+
+    @Override
+    @Transactional
+    public CommandProcessingResult createInterestPause(ExternalId loanExternalId, LocalDate startDate, LocalDate endDate, String dateFormat,
+            String locale) {
+        this.context.authenticatedUser();
+
+        Long loanId = loanRepository.findIdByExternalId(loanExternalId);
+        if (loanId == null) {
+            throw new LoanNotFoundException("Loan not found with External ID: " + loanExternalId);
+        }
+
+        Loan loan = loanRepository.findByExternalId(loanExternalId).orElseThrow(() -> new LoanNotFoundException(loanExternalId));
+
+        validateInterestPauseDates(loan, startDate, endDate, dateFormat, locale);
+
+        LoanTermVariations variation = new LoanTermVariations(LoanTermVariationType.INTEREST_PAUSE.getValue(), startDate, null, endDate,
+                false, loan);
+
+        LoanTermVariations savedVariation = loanTermVariationsRepository.saveAndFlush(variation);
+
+        return new CommandProcessingResultBuilder().withEntityId(savedVariation.getId()).build();
+    }
+
+    @Override
+    @Transactional
+    public CommandProcessingResult createInterestPauseByLoanId(Long loanId, LocalDate startDate, LocalDate endDate, String dateFormat,
+            String locale) {
+        this.context.authenticatedUser();
+
+        final Loan loan = loanRepository.findById(loanId).orElseThrow(() -> new LoanNotFoundException("Loan not found with ID: " + loanId));
+
+        validateInterestPauseDates(loan, startDate, endDate, dateFormat, locale);
```

Comment:
You can extract this logic into a generic solution and keep the difference minimal:
- fetching loan by id or fetching loan by external id

but the rest is the same in both cases!

Reply:
Fixed